Feat: LSP Type Hints

poliorcetics opened the pull request:
Add support for LSP type hints (merged in version 3.17 of the spec, a few month old now).

For now, I only added it in the main editor views, pickers do not have type hints.
Even if we decide they need to, I would prefer doing it in another PR since this one is already big
enough as it is.

I'll join a screen in a comment below, and if people could try it out with their LSPs and configs, it
would be very useful, since I pretty much only use Rust-Analyzer so testing may be somewhat limited still.

Closes #2070

EDIT: [instructions to test](https://github.com/helix-editor/helix/pull/5934#issuecomment-1445040945)

poliorcetics commented:
![Screenshot 2023-02-12 at 04 12 20](https://user-images.githubusercontent.com/7951708/218290973-43fc17d5-bba1-4c44-9221-b0927ada7146.png)

Proful commented:
Amazing!

![image](https://user-images.githubusercontent.com/354596/218293038-eef64d6d-680f-4b06-8722-f3dfabbdab49.png)

pascalkuthe commented on helix-term/src/commands/lsp.rs:
```diff
@@ -1324,3 +1331,191 @@ pub fn select_references_to_symbol_under_cursor(cx: &mut Context) {
+            let first_visible_line = view.estimate_last_doc_line(doc).saturating_sub(view_height);
+            let first_line = first_visible_line
+                .saturating_sub(view_height)
+                .min(len_lines);
```
the `min` is not necessary since you are subtracting

poliorcetics replied:
Fixed

pascalkuthe commented on helix-term/src/commands/lsp.rs:
```diff
@@ -1324,3 +1331,191 @@ pub fn select_references_to_symbol_under_cursor(cx: &mut Context) {
+            let first_line = first_visible_line
+                .saturating_sub(view_height)
+                .min(len_lines);
+            let last_line = first_visible_line
+                .saturating_add(view_height)
```
you just calculated `last_line - height + height` here. What you want is `last_line + height`.
Also you likely want to calculate `first_line` directly to make this more readable so this should really be `first_line.saturating_add(2*height)`.

poliorcetics replied:
Fixed

pascalkuthe commented on helix-term/src/commands/lsp.rs:
```diff
@@ -1324,3 +1331,191 @@ pub fn select_references_to_symbol_under_cursor(cx: &mut Context) {
+            // than computing all the hints for the full file (which could be dozens of time
+            // longer than the view is).
+            let view_height = view.inner_height();
+            let first_visible_line = view.estimate_last_doc_line(doc).saturating_sub(view_height);
```
This is a bit backwards as `estimate_last_doc_line` interally calcualtes the first line and then adds `view_height`.
Instead just use:
```suggestion
            let first_visible_line = doc.text().char_to_line(view.offfset.anchor);
```

poliorcetics replied:
Fixed

pascalkuthe commented on helix-term/src/commands/lsp.rs:
```diff
@@ -1324,3 +1331,191 @@ pub fn select_references_to_symbol_under_cursor(cx: &mut Context) {
+            let first_char_in_range = doc_slice.line_to_char(first_line);
+            let last_char_in_range = doc_slice.line_to_char(last_line);
+
+            let range = helix_lsp::util::range_to_lsp_range(
+                doc_text,
+                helix_core::Range::new(first_char_in_range, last_char_in_range),
+                language_server.offset_encoding(),
+            );
+
+            (
+                language_server.text_document_range_inlay_hints(doc.identifier(), range, None),
```
I think it would be a bit cleaner if we passed a `helix_core::Range` to `text_document_range_inlay_hints` and do the lsp conversion inside `text_document_range_inlay_hints` in `helix-lsp`

poliorcetics replied:
Eh, I don't like that: for one, it's not the style used by the other functions in this file, and it conflates `lsp::Range` and `helix_core::Range` in place where only LSP-related stuff happens, I don't think that's good for clarity compared to the single conversion call later in the crate chain

pascalkuthe commented on helix-term/src/commands/lsp.rs:
```diff
@@ -1324,3 +1331,191 @@ pub fn select_references_to_symbol_under_cursor(cx: &mut Context) {
+            let view = match editor.tree.try_get_mut(view_id) {
+                Some(view) => view,
+                None => {
+                    // If the view doesn't exist anymore, purge its data from the text annotations
+                    // in the document
+                    doc.reset_text_annotations(view_id);
+                    return;
+                }
```
this should be handled in 

`Document::remove_view`.

`Document::ensure_view_init` might also be relevant to this PR

poliorcetics replied:
Done

pascalkuthe commented on helix-term/src/commands/lsp.rs:
```diff
@@ -1324,3 +1331,191 @@ pub fn select_references_to_symbol_under_cursor(cx: &mut Context) {
+                    );
+                }
+
+                annotations.add_inline_annotations(
+                    Rc::new([InlineAnnotation::new(char_idx, label)]),
+                    style.map(helix_core::syntax::Highlight),
+                );
```
layers carray a certain amount of overhead. Instead of creating a seperate layer for the each annotation create a `Vec<InlineAnnotation>` and then convert that to `Rc<[InlineAnnotation]>` (same for everwhere else you create a layer, there should be only one layer per type of virtual text, one for padding, one for each kind of label (paramters,  types and default style))

poliorcetics replied:
Done

pascalkuthe reviewed (changes requested):
Nice to finally see the virtual text API put to use :tada: 

Sadly I think you missunderstood the API somewhat and therefore had to do a bunch of unnecesassary changes. I designed the virtual text API very carefully (including with this usecase in mind) and the way you changed the API here makes it less efficent and less elegant.

The API is a bit abstract without an example and you are not the first person to be initially confused.  I think the core of your misunderstanding is that you created a single layer (call to `add_inline_annotations`) per inlay a hint

That are a lot of layers (and not efficent since every layers must be checked at every singel char when transvcersing the document) and I therefore you probably thaught that you needed to store these somewhere.

Instead the idea is that one layer corresponds to a single TYPE of virtual text. 
Therefore it is possible to construct these layers on the fly in the `text_annotations` functions and only store the contents of the layers. In your case there should only be the need for 4 layers. Constructing a `Vec` of 4 elements on the fly is basically free and is how the API was supposed to be used.

I left a bunch of comments with more details.

poliorcetics commented on helix-view/src/document.rs:
```diff
@@ -107,6 +107,7 @@ pub struct Document {
     pub(crate) id: DocumentId,
     text: Rope,
     selections: HashMap<ViewId, Selection>,
+    text_annotations: HashMap<ViewId, TextAnnotations>,
```
Done

poliorcetics commented on helix-core/src/text_annotations.rs:
```diff
@@ -154,12 +157,42 @@ pub struct TextAnnotations {
+                apply_change(&annot.anchor_char_idx);
+            }
+        }
+    }
```
Done

poliorcetics commented on helix-view/src/view.rs:
```diff
@@ -125,6 +125,22 @@ pub struct View {
+    pub first_line: usize,
+    /// Last line of the range for which the annotations were computed
+    pub last_line: usize,
 }
```
Done

poliorcetics commented on helix-view/src/document.rs:
```diff
@@ -1254,10 +1286,6 @@ impl Document {
         }
     }
-
-    pub fn text_annotations(&self, _theme: Option<&Theme>) -> TextAnnotations {
```
Done, I still need the `view_id` though, is that a problem ?

pascalkuthe replied:
You can move this entire function into `helix_view::View::text_annotations`.
The intentation was:
* Virtual text that is shared among all views of a buffer is added in `Document::text_annotations`
* Virtual text that is specific to a single view is added in `View::text_annotations`

You have full access to the document in `View::text_annotations` so everything you can do here can be done there too (in fact that function just calls this function).

poliorcetics commented on helix-view/src/document.rs:
```diff
@@ -1154,10 +1177,19 @@ impl Document {
         &self.selections[&view_id]
     }
 
+    #[inline]
```
Made in another commit instead, it's just a small drive-by fix

poliorcetics commented on helix-view/src/view.rs:
```diff
@@ -402,9 +419,9 @@ impl View {
         // TODO custom annotations for custom views like side by side diffs
-        doc.text_annotations(theme)
+        doc.text_annotations(self.id)
     }
```
Fixed

poliorcetics commented:
> Sadly I think you missunderstood the API somewhat and therefore had to do a bunch of unnecesassary changes. I designed the virtual text API very carefully (including with this usecase in mind) and the way you changed the API here makes it less efficent and less elegant.
> 
> The API is a bit abstract without an example and you are not the first person to be initially confused. I think the core of your misunderstanding is that you created a single layer (call to add_inline_annotations) per inlay a hint
>
> That are a lot of layers (and not efficent since every layers must be checked at every singel char when transvcersing the document) and I therefore you probably thaught that you needed to store these somewhere.

Yes, this is really confusing at first because there is pretty much zero doc in Helix about virtual text and annotations (or, if there is, it's absolutely not where I expect it to be, on the types and methods) so I was really confused, thanks for clarifying !

> Instead the idea is that one layer corresponds to a single TYPE of virtual text.
> Therefore it is possible to construct these layers on the fly in the text_annotations functions and only store the contents of the layers. In your case there should only be the need for 4 layers. Constructing a Vec of 4 elements on the fly is basically free and is how the API was supposed to be used.

Yes, that approach is much better, I used it instead. I can't separate padding from the hints though, else it breaks in all kinds of ways: everything is squished together and unreadable.

> Once you apply my suggestion the diff should reduce significantly and this should turn into a smaller PR.

Not really 😅 I added quite some docs for inlay hints in documents that re-inflated the count after reducing it, but still the MR is better than it was before thanks to you

pascalkuthe commented on helix-view/src/view.rs:
```diff
@@ -402,9 +402,10 @@ impl View {
     pub fn text_annotations(&self, doc: &Document, theme: Option<&Theme>) -> TextAnnotations {
         // TODO custom annotations for custom views like side by side diffs
-        doc.text_annotations(theme)
+        doc.text_annotations(self.id, theme)
```
just move all your code here and replace `self` with `doc`

pascalkuthe commented on helix-term/src/commands/lsp.rs:
```diff
@@ -1324,3 +1330,170 @@ pub fn select_references_to_symbol_under_cursor(cx: &mut Context) {
+                    _ => &mut other_inlay_hints,
+                };
+
+                if let Some(true) = hint.padding_left {
+                    inlay_hints_vec.push(InlineAnnotation::new(char_idx, " "));
+                }
```
From the LSP spec:

> Note: Padding should use the editor's background color, not the
> background color of the hint itself. That means padding can be used
> to visually align/separate an inlay hint.

That means that you need one extra layer for the the padding which uses `ui.virtual.whitespace` (the theme key normally used for spaces)

poliorcetics replied:
Fixed, though I introduced `ui.virtual.inlay-hint.padding` instead, to help people have a different style for them than regular whitespaces, else it quickly becomes confusing because there appears to be an unreachable whitespace that is not an inlay hint.

pascalkuthe commented:
> Yes, that approach is much better, I used it instead. I can't separate padding from the hints though, else it breaks in all kinds of ways: everything is squished together and unreadable.
> 

Huh that is odd, it shouldn't let me check.

> Not really sweat_smile I added quite some docs for inlay hints in documents that re-inflated the count after reducing it, but still the MR is better than it was before thanks to you


Well you added quite a bit of docs and its still almost 25% smaller so I would call that a win :D

pascalkuthe commented:
> > Yes, that approach is much better, I used it instead. I can't separate padding from the hints though, else it breaks in all kinds of ways: everything is squished together and unreadable.
> 
> Huh that is odd, it shouldn't let me check.
> 
> > Not really sweat_smile I added quite some docs for inlay hints in documents that re-inflated the count after reducing it, but still the MR is better than it was before thanks to you
> 
> Well you added quite a bit of docs and its still almost 25% smaller so I would call that a win :D

I just implemented this myself and it worked fine for me.
However for this to work the order of layer is important, since all these annotations will occur at the same position:

* first add the left padding layer
* the the actual label layers
* then the right padding layer

pascalkuthe commented on helix-view/src/document.rs:
```diff
@@ -150,13 +156,63 @@ pub struct Document {
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub struct DocumentInlayHintsId {
+    /// Document revision at which the inlay hints were requested.
+    pub revision: usize,
```
I tested out this PR some more and noticed that using the revision for refreshing doesn't work as well as I had hoped (it cases issues with undo for example). It seemed like a good solution at first but I cameup with something better/easier: Since all inline annotations for all views need to be requested on idle timeout when the document is modified simply store a bool on the document:
``` rust
inlayhint_outdated: bool
```

* In `apply_impl` simply set this to ` true`.
* In `compute_inlay_hints_for_view`: reset this field to `false` and rerequest the inlay hints if this is true OR the view position changed

pascalkuthe commented on helix-view/src/document.rs:
```diff
@@ -859,24 +929,46 @@ impl Document {
+            let apply_inlay_hint_changes = |annotations: &mut Rc<[InlineAnnotation]>| {
+                if let Some(data) = Rc::get_mut(annotations) {
+                    for inline in data.iter_mut() {
+                        inline.char_idx = changes.map_pos(inline.char_idx, Assoc::After);
```
Neither `Assoc::After` or `Assoc::Before` works well here that is because both of these were designed for **ranges** where the start is mapped with `Assoc::Before` and `Assoc::After`. However for inlay hints we want something different. Inlay hints are bound to a single point so when that point is removed when also want to delete the inlay hint. For that you need to add a new variant `Assoc::Point` which returns `None` in case the position which contains `char_idx` was removed/replaced.

For that to worth tough we need a way to remove `InlineAnnotations` so the equivalent for `Vec::retain` but for `Rc<[T]>` which is not possible without a custom reference counted type (which we probably don't want to add here).

I am considering just switching `TextAnnotations` back to using slices instead of `Rc`  which it originally did. I changed it to make the API easier to use/remove lifetime issues that occured in #5340 but I think those issues were actually resolved by something else so maybe a slice is a better choice afterall.

I don't have more time to look into this right now but you could start with implementing `Assoc::Delete` and I will try to comeup with something in the next couple of days.

poliorcetics replied:
I'm not sure I want to do this in this PR. Nor that I am qualified to do it.

`ChangeSet::map_pos()` currently unconditionnally returns a `usize`, adding `Assoc::Delete` or `Assoc::Point` would need to change the return type to account for the deletion cleanly.

Instead I think this PR should either be blocked until `map_pos` and `Assoc::...` are better equipped or, if we don't want to block too long, make an issue for it and fix it later on since for now `Assoc::After` works well enough (it probably has strange edge cases at the end of files but apart from that it should be okay for a v1)

pascalkuthe replied:
If you don't feel confident working on this that's fine. I will create a PR myself in the next couple days that that you can then depend on. It's not that big of a change so I dlunt it will block this PR for long.

I think it is a requirement for this because without properly handeling this some odd effects do occur. For example if you delete two lines `2xd` then all inlay hints within those lines are moved to the start of the next line

pascalkuthe replied:
I actually changed my mind on this. Mapping points doesn't work well in general. The right fix here would be to convert these points to ranges by expanding to the word at the specific position and then tracking that range (only removing it when empty just like we do for diagnostics).

I have a pretty good idea how to implement this but it's a bit more effort and I don't have the necessary time right now. In practice this not as large of a problem as I orginally tough because the type hints get refreshed relatively quickly (on idle timeout) so the artificats usually don't stick around for that long. So I think I can do the necessary refactors in a followup PR.

pascalkuthe commented on helix-term/src/ui/editor.rs:
```diff
@@ -978,6 +978,8 @@ impl EditorView {
     }
 
     pub fn handle_idle_timeout(&mut self, cx: &mut commands::Context) -> EventResult {
+        commands::compute_inlay_hints_for_all_views(cx.editor, cx.jobs);
```
I am not 100% sure why but the inlay hints don't seem to update for me while scrolling with the mouse.
It seems that we do not reset the idle timeout on mouse actions. I think that might actually be an existing bug that would be fixed by adding:
``` rust
cx.editor.reset_idle_timer();
```
at the top of `handle_mouse_event`

archseer commented on helix-view/src/document.rs:
```diff
@@ -1255,8 +1348,58 @@ impl Document {
+    }
+
+    /// Get the inlay hints for this document and `view_id`.
+    pub fn get_inlay_hints(&self, view_id: ViewId) -> Option<&DocumentInlayHints> {
```
Nit: I usually use `inlay_hints()` (no `get_`) on the getters

archseer commented on helix-view/src/document.rs:
```diff
@@ -150,13 +156,63 @@ pub struct Document {
+    pub id: DocumentInlayHintsId,
+
+    /// Inlay hints of `TYPE` kind, if any.
+    pub type_inlay_hints: Rc<[InlineAnnotation]>,
```
We should probably be using `Arc` everywhere. I've avoided `Rc`

pascalkuthe replied:
In this case the only reasons this is a `Rc` instead of a `Vec` is that the virtual text API currently expects that instead of a slice. There are some other limitations imposed by using an Rc so I might just switch backto a slice, see https://github.com/helix-editor/helix/pull/5934#discussion_r1103890555.

I think there were some lifetime confusions initially in #5340 but I think those were resolved otherwise and I just kept the Rc to make the API easier to use.

poliorcetics replied:
Resolving in favor of https://github.com/helix-editor/helix/pull/5934#discussion_r1103890555, easier for me to track one comment than two

archseer commented on helix-term/src/commands/lsp.rs:
```diff
@@ -1324,3 +1330,170 @@ pub fn select_references_to_symbol_under_cursor(cx: &mut Context) {
 }
+
+pub fn compute_inlay_hints_for_all_views(editor: &mut Editor, jobs: &mut crate::job::Jobs) {
+    if !editor.config().lsp.display_inlay_hints {
+        for doc in editor.documents.values_mut() {
+            doc.reset_all_inlay_hints();
+        }
+        return;
+    }
```
Could we avoid doing this here and instead do the reset once when editor config is updated?

poliorcetics replied:
Yes, it's better, done

poliorcetics reviewed (commented):
Fixed some of the comments, I'm gonna work on the bigger ones now

lazytanuki commented:
Hi and thanks for working on this!
Been testing for a couple hours and it's really nice.
I find that having the hints pop up in the function arguments while typing them can be a bit unnerving, as it shifts the cursor back and forth.
I think it would be nice to have a setting to disable inlay hints for the current function parameters, in insert mode a least. Or maybe in insert mode altogether ? What do others think ?

poliorcetics commented:
Those should probably be options. I agree that some of that sound nice, though I think implementing them in this MR would delay it too much: I would prefer having something working merged then refine it as user reports pop up :), that way we can both keep this MR relatively small (my time is sadly limited, the maintainers' time too) and account for lots of reports at once instead of just the people following this MR

Does that sound good to you ?

univerz commented:
>  I find that having the hints pop up in the function arguments while typing them can be a bit unnerving

+1. I didn't notice this problem while using vscode, so there is a way to do it right.

marcbull commented:
Thanks for the work on this part of Helix, this is one of the features I really want to use.

> I find that having the hints pop up in the function arguments while typing them can be a bit unnerving, as it shifts the cursor back and forth.

If I compare it with VS code, the experience there is more pleasant because there is a delay after the last change until the LSP type hint is adjusted.

I don't think it would be good to disable the LSP type hint in the current line or completely in insert mode, because I personally like to check the result value as I type and it's a welcome hint to see if the result is an option<T> or already a T, especially if you're new to a project or using a new API.

poliorcetics commented:
> I find that having the hints pop up in the function arguments while typing them can be a bit unnerving, as it shifts the cursor back and forth.

I think part of the issue is that the default `idle-timeout` of helix is really high, can you try with the config below to see if it improves the situation (I expect it will still happen, just less visibly) ? (Test it once the LSP is fully started, on file opening there'll always be a delay for LSP startup)

```toml
[editor]
idle-timeout = 8 # for 120 fps, use 16 for 60
```

marcbull commented:
> I think part of the issue is that the default `idle-timeout` of helix is really high, can you try with the config below to see if it improves the situation (I expect it will still happen, just less visibly) ? (Test it once the LSP is fully started, on file opening there'll always be a delay for LSP startup)
> 
> ```toml
> [editor]
> idle-timeout = 8 # for 120 fps, use 16 for 60
> ```

For my part, I can say that with a value of 120 (equivalent to 2 seconds) my expectations are met and it behaves as in VS Code. During the input, it then no longer changes erratically.

pascalkuthe commented:
> > I find that having the hints pop up in the function arguments while typing them can be a bit unnerving, as it shifts the cursor back and forth.
> 
> I think part of the issue is that the default `idle-timeout` of helix is really high, can you try with the config below to see if it improves the situation (I expect it will still happen, just less visibly) ? (Test it once the LSP is fully started, on file opening there'll always be a delay for LSP startup)
> 
> ```toml
> [editor]
> idle-timeout = 8 # for 120 fps, use 16 for 60
> ```

Please don't reduce the idle timeout to such low values. It is known to cause issues and unnecessary overhead. Nothing below 100(ms) should be necessary.

lazytanuki commented:
I have my `idle-timeout` at 100, I don't really want to have it at 2s just for this feature. Here's an example of what I am talking about:

https://user-images.githubusercontent.com/43273245/219120731-6eaa611b-39e3-41c2-8beb-a726d0cb4df9.mp4

> I don't think it would be good to disable the LSP type hint in the current line or completely in insert mode

I was suggesting a config option here. However, I don't know what the correct defaults should be.

marcbull commented:
I agree that you should not use idle time for this. It should only be a temporary change to test the effects. I think a separate timeout for the LSP type hint is a good solution.

pascalkuthe commented on helix-term/src/ui/editor.rs:
```diff
@@ -1002,6 +1004,8 @@ impl EditorView {
         event: &MouseEvent,
         cxt: &mut commands::Context,
     ) -> EventResult {
+        cxt.editor.reset_idle_timer();
```
```suggestion
	if event.kind != MouseEventKind::Moved{
		cxt.editor.reset_idle_timer();
	}
```

I don't think we should be triggering the idle timeout if the cursor is just moved

pascalkuthe commented on helix-view/src/editor.rs:
```diff
@@ -386,6 +388,7 @@ impl Default for LspConfig {
             display_messages: false,
             auto_signature_help: true,
             display_signature_help_docs: true,
+            display_inlay_hints: true,
```
```suggestion
            display_inlay_hints: false,
```

same here

pascalkuthe commented on book/src/configuration.md:
```diff
@@ -118,9 +118,11 @@ The following statusline elements can be configured:
 | `enable`              | Enables LSP integration. Setting to false will completely disable language servers regardless of language settings.| `true` |
 | `display-messages`    | Display LSP progress messages below statusline[^1]          | `false` |
 | `auto-signature-help` | Enable automatic popup of signature help (parameter hints)  | `true`  |
+| `display-inlay-hints` | Display inlay hints[^2]                                     | `true`  |
```
```suggestion
| `display-inlay-hints` | Display inlay hints[^2]                                     | `false`  |
```

For now I don't think we should enable this by default until the implementation is improved further from this MVP.

pascalkuthe reviewed (changes requested):
I actually think that using the idle timeout for a MVP is fine.

The correct solution (what VSCode  does) would be to have a custom background task that responds to:
* any document modified (if visible)  with a long debounce timeout (3s)
* view moved/focused event with a short debounce timeout (100ms)

This ensures that:
* responses to edits are not too fast
* changes in other splits that affect inlay hints in the current document are also picked up (type check/name resolution are global analysis after all)
* response to scrolling is very fast to reduce noticeable pop-in when possible (while still avoiding excessive requests with debouncing)


Implementing that will require some refactors and is more tricky than just using idle timeout. The implementation would be similar to the diff gutters in some ways (although quite a bit simpler).
I don't think the issues here are bad enough to block on that tough. I can handle that in a followup PR. Implementing this will require centralizing the modifications of the view positions (so we can hook into that) which is something I want to do for the diffmode anyway.

For now I think we should not enable the feature by default. I also added one small suggestion otherwise this LGTM

poliorcetics commented:
Fixed your comments and added a note for users so that they are warned and can report any jankiness in inlay hints, so we can improve them where needed.

Thanks for the help and the review, it was really helpful!

poliorcetics commented:
<img width="884" alt="In the line telling a force push happened, the words 'force pushed' are a link to the diff from just before the force push to just atfer" src="https://user-images.githubusercontent.com/7951708/219199846-a31e4dc8-ae08-4765-bd28-814e8ccdb37b.png">

By the way, Github tip that you may not know: to review diffs in force-push, `force pushed` is a link in the screen above

pascalkuthe reviewed (approved):
LGTM :+1: 

Thanks for tackling this! It's a cool showcase for what the new virtual text support from #5420. I wanted to add this myself eventually but as I am quite busy right now I am happy you took that over for me. Despite needing some further enhancements this MVP already works quite well.

pascalkuthe commented on helix-view/src/view.rs:
```diff
@@ -402,9 +404,53 @@ impl View {
+        let other_style = theme
+            .and_then(|t| t.find_scope_index("ui.virtual.inlay-hint"))
+            .map(Highlight);
+        let padding_style = theme
+            .and_then(|t| t.find_scope_index("ui.virtual.inlay-hint.padding"))
+            .map(Highlight);
```
Ah one small think I forgot to add a review comment for:

I don't think this highlight scope should exist.
The LSP spec specifically states that the padding should have the same background color as the rest of the editor and NOT the same background as the virtual text.

Because of the way theme scopes work this will fallback to `ui.virtual.inlay-hint` if a theme doesn't define `ui.virtual.inlay-hint.padding` and therefore would not work as the LSP spec intended. I don't think it's necessary to style this empty space at all since the LSP spec is very specific that it should have the same background color as normal text.

You can simply use `None` here (in that case `ui.text` will be automatically used) :+1: 
```suggestion
        let padding_style = None;
```

poliorcetics replied:
I don't agree with the spec here, I find it confusing in monospaced text to have a regular space I can't reach with my cursor, but done, you're right we should follow the spec regardless

jquesada2016 commented:
What are the steps needed to test this out in Rust? Do I need to add anything to `languages.toml`? Or is this automatic?

I tried both with and without changes, with the changes listed below, and did not see inlay hints.


```toml
# [[language]]
# name = "rust"
#
# [language.config.hints]
# assignVariableTypes = true
# compositeLiteralFields = true
# constantValues = true
# functionTypeParameters = true
# parameterNames = true
# rangeVariableTypes = true
```

I saw thins config on a different issue, but not sure if these apply.

poliorcetics commented:
@jquesada2016 You'll need to activate the option in the `~/.config/helix/config.toml`:

```toml
[editor.lsp]
display-inlay-hints = true
```

and then use the parameters documented by rust-analyzer to activate what you want amongst inlay hints: https://rust-analyzer.github.io/manual.html#configuration (search for `inlay`)

As an example, here's what I use:

```toml
[[rust]]

[language.config]
inlayHints.bindingModeHints.enable = false
inlayHints.closingBraceHints.minLines = 10
inlayHints.closureReturnTypeHints.enable = "with_block"
inlayHints.discriminantHints.enable = "fieldless"
inlayHints.lifetimeElisionHints.enable = "skip_trivial"
inlayHints.typeHints.hideClosureInitialization = false
```

jquesada2016 commented:
Awesome! That worked! Thank you very much.

Would you happen to know what theme options affectt it?

poliorcetics commented:
I updated the comment above

jquesada2016 commented:
Thank you very much. This has been THE feature I've been longing for in helix! I work so much faster thanks to inlay hints. I can finally switch to this editor full-time!

David-Else commented:
It is working great, thanks!

Is it possible to style the hints like VS Code (without the rounded edges I assume)?:

![inlay-hints](https://user-images.githubusercontent.com/12832280/221609262-16c60118-3b17-44b7-8c91-bac25c7baa14.png)

Dark Plus styles them in white at the moment as it has no `"ui.virtual.inlay-hint" ` yet, and that doesn't look great as they really need to either stand out or be a lot fainter so they are obviously different from the rest of the code:

![Screenshot from 2023-02-27 15-43-48](https://user-images.githubusercontent.com/12832280/221610288-20049d9d-a854-4b9a-a846-2c6787dcfc6a.png)

pascalkuthe commented:
> It is working great, thanks!
> 
> Is it possible to style the hints like VS Code (without the rounded edges I assume)?:
> 
> ![inlay-hints](https://user-images.githubusercontent.com/12832280/221609262-16c60118-3b17-44b7-8c91-bac25c7baa14.png)
> 
> Dark Plus styles them in white at the moment as it has no `"ui.virtual.inlay-hint" ` yet, and that doesn't look great as they really need to either stand out or be a lot fainter so they are obviously different from the rest of the code:
> 
> ![Screenshot from 2023-02-27 15-43-48](https://user-images.githubusercontent.com/12832280/221610288-20049d9d-a854-4b9a-a846-2c6787dcfc6a.png)

We don't require theme adjustements for merging, those are generally done in seperate PRs (unless @poliorcetics explicitly wants to add those themselves) just like for diff gutters

David-Else commented:
@pascalkuthe My question was more general, I don't know how to make the text have a different background color than the actual main background (like in the VS Code example). I would like to try and match it for Dark Plus and VS Code if it is possible.

pascalkuthe commented:
> @pascalkuthe My question was more general, I don't know how to make the text have a different background color than the actual main background (like in the VS Code example). I would like to try and match it for Dark Plus and VS Code if it is possible.

you should just be able to set the `bg` color so: `"ui.virtual.inlay-hint" = {fg = ..., bg = ... }`. You can in theorey do this for all test highlights but those will be overwritten by selection but that is not the case for inlay hints as they can't be selected.

Note that inlay hint background color (if set) does overwrite the cursorline color. It's a matter of taste what should happen in that case. In my opinion the current behavior is fine but it's something to be aware of while theming

evanrichter commented:
> When might this PR get merged?

when it's ready. if you want to try it now, you can build my fork: https://github.com/evanrichter/helix

David-Else commented:
@KittyBorgX https://github.com/helix-editor/helix/discussions/5883

swarnimarun commented:
Nice, I tried a few things and will use it for work for a bit if I find any issues I will let folks here know, in case it isn't merged until then.

What my theme and setup look like atm,

![some-rust-code](https://user-images.githubusercontent.com/19930870/221717913-ba2e4d93-170f-4fe8-ba67-901c4c7e4fc2.png)

davxy commented:
- Does this works fine for other languages as well (other than Rust). If yes, which one? Any experience to share?
- Is there an option + keybind to toggle inline hints when are enabled? Sometimes I want a quick hint but then I may want to work without the hints

paholg commented:
>     * Is there an option + keybind to toggle inline hints when are enabled? Sometimes I want a quick hint but then I may want to work without the hints

You can set a keybind to `:toggle-option lsp.display-inlay-hints`.

pascalkuthe commented:
This works for languages whose LS supports inlay hints. The feature is pretty new in the LSP standard so only to most up to date version of actively developed LS will support it

makiri1993 commented:
Was somebody able to get it working with typescript-lsp? Rust worked pretty much without config and very nice 🙌 

My config for ts:

```
[[language]]
name = "typescript"
formatter = { command = "prettier", args = ["--parser", "typescript"] }
auto-format = true
max-line-length = 100
config = { inlayHints = { includeInlayParameterNameHints = "all", includeInlayParameterNameHintsWhenArgumentMatchesName = true, includeInlayFunctionParameterTypeHints = true, includeInlayVariableTypeHints = true, includeInlayVariableTypeHintsWhenTypeMatchesName = true, includeInlayPropertyDeclarationTypeHints = true, includeInlayFunctionLikeReturnTypeHints = true, includeInlayEnumMemberValueHints = true } }


```

erasin commented:
@makiri1993 https://github.com/helix-editor/helix/issues/2070#issuecomment-1433195804

The config for `golang`, `lua`, `typescript/js`.


```toml
# language.toml config for inlay hit

[[language]]
name = "go"

[language.config.hints]
assignVariableTypes = true
compositeLiteralFields = true
constantValues = true
functionTypeParameters = true
parameterNames = true
rangeVariableTypes = true

[[language]]
name = "lua"

[language.config.Lua.hint]
enable = true
arrayIndex = "Enable"
setType = true
paramName = "All"
paramType = true
await = true


[[language]]
name = "typescript"

[language.config]
hostInfo = "helix"

[language.config.typescript.inlayHints]
includeInlayEnumMemberValueHints = true
includeInlayFunctionLikeReturnTypeHints = true
includeInlayFunctionParameterTypeHints = true
includeInlayParameterNameHints = "all"                       # 'none' | 'literals' | 'all';
includeInlayParameterNameHintsWhenArgumentMatchesName = true
includeInlayPropertyDeclarationTypeHints = true
includeInlayVariableTypeHints = true

[[language]]
name = "tsx"

[language.config]
hostInfo = "helix"

[language.config.typescript.inlayHints]
includeInlayEnumMemberValueHints = true
includeInlayFunctionLikeReturnTypeHints = true
includeInlayFunctionParameterTypeHints = true
includeInlayParameterNameHints = "all"                       # 'none' | 'literals' | 'all';
includeInlayParameterNameHintsWhenArgumentMatchesName = true
includeInlayPropertyDeclarationTypeHints = true
includeInlayVariableTypeHints = true

[[language]]
name = "javascript"

[language.config]
hostInfo = "helix"

[language.config.javascript.inlayHints]
includeInlayEnumMemberValueHints = true
includeInlayFunctionLikeReturnTypeHints = true
includeInlayFunctionParameterTypeHints = true
includeInlayParameterNameHints = "all"                       # 'none' | 'literals' | 'all';
includeInlayParameterNameHintsWhenArgumentMatchesName = true
includeInlayPropertyDeclarationTypeHints = true
includeInlayVariableTypeHints = true

[[language]]
name = "jsx"

[language.config]
hostInfo = "helix"
locale = "zh-CN" 

[language.config.javascript.inlayHints]
includeInlayEnumMemberValueHints = true
includeInlayFunctionLikeReturnTypeHints = true
includeInlayFunctionParameterTypeHints = true
includeInlayParameterNameHints = "all"                       # 'none' | 'literals' | 'all';
includeInlayParameterNameHintsWhenArgumentMatchesName = true
includeInlayPropertyDeclarationTypeHints = true
includeInlayVariableTypeHints = true
```

ref:

Golang https://github.com/golang/tools/blob/master/gopls/doc/settings.md#inlayhint
Lua https://github.com/LuaLS/lua-language-server/wiki/Settings#hint
Typescript https://github.com/typescript-language-server/typescript-language-server#inlay-hints-textdocumentinlayhint

xiantang commented:
https://github.com/neovim/neovim/pull/20130 related

archseer commented:
> [neovim/neovim#20130](https://github.com/neovim/neovim/pull/20130) related

We actually already have this implemented. https://github.com/helix-editor/helix/pull/5420

archseer commented:
@poliorcetics can we add the Rust config you provided + @erasin's config to the defaults? Then this PR should be good to merge 👍🏻

poliorcetics commented:
Rebased and configs added

pascalkuthe commented:
Yeah bacoee behaves the same and it's very tricky to change because we do selectionswith syntax highlights. If we applied those to virtual text than virtual text would have the full style of the text (so syntax highlight color, dagnowtic underlines...)

erasin commented:
there has a problem that the cursor position in insert mode is after the virtual text.

helix:
![hx](https://user-images.githubusercontent.com/716514/224053964-e173e9f8-5f1e-4fe4-89e3-2b665c8aa4b7.png)

vscode:
![vs](https://user-images.githubusercontent.com/716514/224053990-079ac268-822d-4cbb-8de8-e0b00c2a1056.png)

poliorcetics commented:
I can’t do anything about that, all inlay hints in Helix will behave like that due to how they’re rendered, though I agree it isn’t the most agreable.

Writing a « dumb » fix here (offsetting by one) for type hints would break for arguments and hints at the end of a line

pascalkuthe commented:
> I've been using this pr for a week or two, and I didn't realize the problem before. There has a problem that the cursor position in insert mode is after the virtual text.
> 
> helix:
> 
> ![hx](https://user-images.githubusercontent.com/716514/224056838-01808fbb-9f9c-47f6-9448-f7b47ee3af35.png)
> 
> vscode:
> 
> ![vs](https://user-images.githubusercontent.com/716514/224056859-c36ce80e-b4d9-417e-979f-b65ebe28253a.png)

It's not perfect but yeah as @poliorcetics said that's kind of just how the cursor works in helix.  In insertmode we increase the selection size by one and insert before the cursor block. If you use block cursors its much more clear why it works the way it does: The selection extends one character after the virtual text. Maybe if we switch to zero width cursors oneday that could be solved but i am not sure about that. I could also do some evil trickery here to sepcical case bar cursors but I would rather not do that.

Lordxan commented:
> > I've been using this pr for a week or two, and I didn't realize the problem before. There has a problem that the cursor position in insert mode is after the virtual text.
> > helix:
> > ![hx](https://user-images.githubusercontent.com/716514/224056838-01808fbb-9f9c-47f6-9448-f7b47ee3af35.png)
> > vscode:
> > ![vs](https://user-images.githubusercontent.com/716514/224056859-c36ce80e-b4d9-417e-979f-b65ebe28253a.png)
> 
> It's not perfect but yeah as @poliorcetics said that's kind of just how the cursor works in helix. In insertmode we increase the selection size by one and insert before the cursor block. If you use block cursors its much more clear why it works the way it does: The selection extends one character after the virtual text. Maybe if we switch to zero width cursors oneday that could be solved but i am not sure about that. I could also do some evil trickery here to sepcical case bar cursors but I would rather not do that.

I actually would like to see an optional patch for that (sorry can't do it myself just started learning rust)

erasin commented:
Since I am a non-ascii characters user, the block cursor is so unfriendly to Chinese characters that I hardly ever use it.

The fact that this problem was not realized before means that it does not affect me much. Hope to become a better experience in the future.

thanks

poliorcetics commented:
(I'm won't do it in this MR but here is a potential solution):

- If the cursor is next to an inlay hint where the format is `text<cursor><space>` or `<space><cursor><text>`, we could display the cursor closer next to the text, adapting the position of the hint locally to obtain: `text<cursor><hint><space>` and `<space><hint><cursor>text`
- In the case of a conflict (`text<hint>text`), we would just defaut to the current behaviour.

I'm not certain how much work it would be to do that, or if it's even possible, but if it works it would probably solve most of the problematic cases ?

archseer commented:
I've noticed this issue too, but given that it's a new feature that's also off by default we can resolve this in a follow up.

A quick solution we could use: only display hints outside of insert mode. That would also address issues with the hints shifting the code around while you type. Not sure how desirable this would be, the hints are helpful when typing out method parameters. Though that's also resolved by signature help. It just seems to me that hints are more beneficial during reading the code (normal mode) than writing (insert mode).

poliorcetics commented:
> only display hints outside of insert mode.

I don't know, things disappearing just before I start typing seems very jarring too.

> It just seems to me that hints are more beneficial during reading the code (normal mode) than writing (insert mode).

I'm also reading code when I type some more, adding code is not done in a vacuum

If we do this it should be an option, because it seems to me to be very much useless to only have type hints outside of insert mode

jquesada2016 commented:
> only display hints outside of insert mode.

I personally think this is a really bad idea. For example, when you're chaining method calls, it's often hard to remember what types the method will return, and being able to see it while in insert mode is very helpful.

erasin commented:
> only display hints outside of insert mode.

Unlike #6059, type hints is very useful in insert mode. We do not need additional option.

archseer commented:
Agreed, it was just an idea.

evanrichter commented:
so hyped for all the fantastic virtual text features coming down the pipeline. the next release is going to be off the charts! thanks to the maintainers and everyone involved <3

Hywan commented:
Awesome!

eulerdisk commented:
I think inlay-hints theme text style should fallback to the comment style.
Now a lot of theme are "broken", inlay hints for types are like user written types.

cidit commented:
do i need to use helix manually compiled on main to use this feature ?

gabydd commented:
No this was released in 23.05 I believe, you just need to enable it in your config. And for some languages you might need to enable it on the actual language server.
